pkg/node/node.go

```diff
@@ -21,23 +21,48 @@ package node
 import (
 	"context"
 	"os"
+	"time"
 
 	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
+var logger = log.Logger{Logger: logf.Log.WithName("Node")}
+
+var nodeRetry = wait.Backoff{
+	Steps:    5,
+	Duration: 5 * time.Second,
+	Factor:   1.2,
+	Jitter:   0.1,
+}
+
 func GetLocalNode(clientset kubernetes.Interface) (*v1.Node, error) {
 	nodeName, ok := os.LookupEnv("NODE_NAME")
 	if !ok {
 		return nil, errors.New("error reading the NODE_NAME from the environment")
 	}
 
-	node, err := clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to find local node %q", nodeName)
-	}
+	var node *v1.Node
+
+	err := retry.OnError(nodeRetry, func(err error) bool {
+		logger.Infof("retrying to read local node details: %s", err)
+		return true
+	}, func() error {
+		var err error
+
+		node, err = clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "unable to find local node %q", nodeName)
+		}
+
+		return nil
+	})
```

Comment:
Would we expect a subsequent retry to succeed? If not, this will try 5 times - after that we'll still return an error and the pod will exit and restart so I'm not sure what this actually solves. Plus there are other places where we access the API server on startup where we don't retry and could similarly cause an exit and restart.

From the summary in the associated issue, it seems the motivation for this is to avoid a possible DP failure due to pod restart. I'm not sure of the details there but it seems we should try to address that (if possible) since restarts could still occur (unless we try to have the whole process retry indefinitely on startup).

Reply:
Well, there are two points here:

1. To avoid unnecessary traffic hit, I agree that it's better if GW pod restarts don't affect DP at all. But in the current implementation, there is a brief DP failure when GW pod is restarted (for libreswan cable driver).

2. We would like to reduce the restart counter for GW pod as much as possible, we have seen the GW restart counter to be ~15K on some deployments recently over a period of 3-4 months, although it seems that root cause was almost always some temporary failure in the k8s-api or some other k8s-infra and not something directly related to GW functionality.

I think it's worth retrying to call the k8s-api again in places where it can reduce the number of restarts caused by a temporary malfunction in the k8s-api even just to address point 2. 

> Plus there are other places where we access the API server on startup where we don't retry and could similarly cause an exit and restart

I think we should address also the other places. could you please share the relevant places?

Reply:
Agree with Yossi. Not all API failures could trigger a restart, at some places in the code we already have the retry mechanism, but this PR addresses and important issue what was seen in some customer deployments. Lets get this merged and look for potential places where we have to retry.

Reply:
OK. It looks like this is the first place in the GW where we access the API server so if this eventually succeeds then subsequent accesses should succeed as well.